File: Stage29/Python/Problem1753.py
import sys
import heapq

input = sys.stdin.readline

def main():
    V, E = map(int, input().split())
    K = int(input())

    graph = [[] for _ in range(V + 1)]

    for _ in range(E):
        u, v, w = map(int, input().split())
        graph[u].append((v, w))

    INF = int(1e9)
    dist = [INF] * (V + 1)

    pq = []
    dist[K] = 0
    heapq.heappush(pq, (0, K))

    while pq:
        d, now = heapq.heappop(pq)

        if d > dist[now]:
            continue

        for next, w in graph[now]:
            cost = dist[now] + w

            if dist[next] > cost:
                dist[next] = cost
                heapq.heappush(pq, (cost, next))

    for i in range(1, V + 1):
        if dist[i] == INF:
            print("INF")
        else:
            print(dist[i])

if __name__ == "__main__":
    main()

File: Stage29/Java/Problem1753.java
package Stage29.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Problem1753 {
    static class Node implements Comparable<Node> {
        int v, w;

        Node(int v, int w) {
            this.v = v;
            this.w = w;
        }

        @Override
        public int compareTo(Node o) {
            return this.w - o.w;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());

        int K = Integer.parseInt(br.readLine());

        ArrayList<Node>[] graph = new ArrayList[V + 1];
        for (int i = 0; i <= V; i++) {
            graph[i] = new ArrayList<>();
        }

        while (E-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            graph[u].add(new Node(v, w));
        }

        int[] dist = new int[V + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);

        PriorityQueue<Node> pq = new PriorityQueue<>();

        dist[K] = 0;
        pq.add(new Node(K, 0));

        while (!pq.isEmpty()) {
            Node cur = pq.poll();

            int now = cur.v;
            int weight = cur.w;

            if (weight > dist[now]) continue;

            for (Node next : graph[now]) {
                if (dist[next.v] > dist[now] + next.w) {
                    dist[next.v] = dist[now] + next.w;
                    pq.add(new Node(next.v, dist[next.v]));
                }
            }
        }
        
        for (int i = 1; i <= V; i++) {
            if (dist[i] == Integer.MAX_VALUE) sb.append("INF\n");
            else sb.append(dist[i]).append("\n");
        }

        System.out.println(sb);
        br.close();
    }
}
